src/Build/BackEnd/BuildManager/BuildManager.cs

```diff
@@ -1895,6 +1896,54 @@ private void CheckForActiveNodesAndCleanUpSubmissions()
             }
         }
 
+        /// <summary>
+        /// Handles creating new nodes.
+        /// </summary>
+        /// <param name="response">Details of how to make the nodes (the number, type, and so on).</param>
+        private void CreateNodes(ScheduleResponse response)
+        {
+            var newNodes = new ConcurrentBag<NodeInfo>();
+            GetNodeConfiguration();
+            try
+            {
+                Parallel.For(0, response.NumberOfNodesToCreate, i =>
```

Comment:
Do you have any idea whether it's common to have multiple nodes requested by the scheduler at the same time, and how much time parallelization would save? If it's the common case, then the extra parallelization complexity might make sense. If not, it might be better to have less complex code.

Reply:
I believe it will be the default on .NET Core. Ultimately, creating a node itself doesn't take too much time; this is more important in the unusual case in which nodes are repeatedly failing right after they're made as in #860. What appeared to be a hang was probably just up to 32 consecutive 30-second timeouts—one per failed node.

I'd be happy to come over and talk about it if you think we might want to go with a change that's easier to get right.

Reply:
I think in #860 you get a true hang, not a timeout apparent hang. If the main node connects to the child, but the child does not write any handshake back, the main node blocks indefinitely on full framework. I see that [on .net core there's a timeout](https://github.com/microsoft/msbuild/blob/0f85cc7476897ff7bb6f099a1b78ab5fff64b50c/src/Build/BackEnd/Components/Communications/NodeProviderOutOfProcBase.cs#L348-L352), so probably full framework does not have the timeout API.

Thinking back on the main problem, maybe parallel creation is not the ideal fix here. If the main node fails to spawn a child node due to a child node corruption, there's a high chance all nodes will fail that way, since the way the main node sets up a child node does not vary. One caveat is that, before the child node writes a handshake, we don't know whether it's corrupted, or whether the machine is so slow the node didn't get to it yet. Maybe a better way is to:
- fail the build if two node creations failed with the same non-timeout issues.
- for slow machines, figure out some better solution. For example maybe temporarily create the node processes with super high priority so the OS scheduler is forced to schedule their process, and then lower their priority right after they've replied their handshake. This way maybe we can do without the timeout, or have a much shorter timeout.